solr/core/src/test/org/apache/solr/rest/TestManagedFileStorage.java

```diff
@@ -0,0 +1,105 @@
+    try (SolrResourceLoader loader = new SolrResourceLoader(instanceDir.toPath())) {
+      NamedList<String> initArgs = new NamedList<>();
+      String managedDir = instanceDir.getAbsolutePath() + File.separator + "managed";
+      initArgs.add(ManagedResourceStorage.STORAGE_DIR_INIT_ARG, managedDir);
+      FileStorageIO fileStorageIO = new FileStorageIO();
+      fileStorageIO.configure(loader, initArgs);
+      TestManagedFileStorage.doStorageTests(loader, fileStorageIO);
+    }
+  }
+
+  /**
+   * Called from tests for each storage type to run creation and update tests on a persisted managed
+   * resource.
+   */
+  @SuppressWarnings("unchecked")
+  public static void doStorageTests(SolrResourceLoader loader, StorageIO storageIO)
+      throws Exception {
+    String resourceId = "/test/foo";
+
+    JsonStorage jsonStorage = new JsonStorage(storageIO, loader);
+
+    Map<String, String> managedInitArgs = new HashMap<>();
+    managedInitArgs.put("ignoreCase", "true");
+    managedInitArgs.put("dontIgnoreCase", "false");
+
+    List<String> managedList = new ArrayList<>(); // we need a mutable List for this test
+    managedList.addAll(Arrays.asList("a", "b", "c", "d", "e"));
+
+    Map<String, Object> toStore = new HashMap<>();
+    toStore.put(ManagedResource.INIT_ARGS_JSON_FIELD, managedInitArgs);
+    toStore.put(ManagedResource.MANAGED_JSON_LIST_FIELD, managedList);
+
+    jsonStorage.store(resourceId, toStore);
+
+    String storedResourceId = jsonStorage.getStoredResourceId(resourceId);
+    assertTrue(storedResourceId + " file not found!", storageIO.exists(storedResourceId));
+
+    Object fromStorage = jsonStorage.load(resourceId);
+    assertNotNull(fromStorage);
+
+    Map<String, Object> storedMap = (Map<String, Object>) fromStorage;
+    Map<String, Object> storedArgs =
+        (Map<String, Object>) storedMap.get(ManagedResource.INIT_ARGS_JSON_FIELD);
+    assertNotNull(storedArgs);
+    assertEquals("true", storedArgs.get("ignoreCase"));
+    List<String> storedList = (List<String>) storedMap.get(ManagedResource.MANAGED_JSON_LIST_FIELD);
+    assertNotNull(storedList);
+    assertEquals(managedList.size(), storedList.size());
+    assertTrue(storedList.contains("a"));
+
+    // now verify you can update existing data
+    managedInitArgs.put("anotherArg", "someValue");
+    managedList.add("f");
+    jsonStorage.store(resourceId, toStore);
+    fromStorage = jsonStorage.load(resourceId);
+    assertNotNull(fromStorage);
+
+    storedMap = (Map<String, Object>) fromStorage;
+    storedArgs = (Map<String, Object>) storedMap.get(ManagedResource.INIT_ARGS_JSON_FIELD);
+    assertNotNull(storedArgs);
+    assertEquals("someValue", storedArgs.get("anotherArg"));
+    storedList = (List<String>) storedMap.get(ManagedResource.MANAGED_JSON_LIST_FIELD);
+    assertNotNull(storedList);
+    assertEquals(managedList.size(), storedList.size());
+    assertTrue(storedList.contains("e"));
```

Comment:
minor/carry-over from existing code: `f` instead of `e` here since the latter was already previously there and only the former is newly added?
```suggestion
    assertTrue(storedList.contains("f"));
```